crates/ere-openvm/src/lib.rs

```diff
@@ -149,34 +149,63 @@ impl EreOpenVM {
             agg_pk,
             agg_vk,
             app_commit,
-            _resource,
+            resource,
         })
     }
 
-    fn sdk(&self) -> Result<Sdk, CommonError> {
-        let sdk =
-            Sdk::new_without_transpiler(self.app_config.clone()).map_err(CommonError::SdkInit)?;
+    fn cpu_sdk(&self) -> Result<CpuSdk, CommonError> {
+        let sdk = CpuSdk::new_without_transpiler(self.app_config.clone())
+            .map_err(CommonError::SdkInit)?;
         let _ = sdk.set_app_pk(self.app_pk.clone());
         let _ = sdk.set_agg_pk(self.agg_pk.clone());
         Ok(sdk)
     }
+
+    #[cfg(feature = "cuda")]
+    fn gpu_sdk(&self) -> Result<openvm_sdk::GpuSdk, CommonError> {
+        let sdk = openvm_sdk::GpuSdk::new_without_transpiler(self.app_config.clone())
+            .map_err(CommonError::SdkInit)?;
+        let _ = sdk.set_app_pk(self.app_pk.clone());
+        let _ = sdk.set_agg_pk(self.agg_pk.clone());
+        Ok(sdk)
+    }
+}
+
+macro_rules! with_sdk {
+    ($zkvm:ident, |$sdk:ident| $f:expr) => {
+        match $zkvm.resource {
+            ProverResourceType::Cpu => {
+                let $sdk = $zkvm.cpu_sdk()?;
+                $f
+            }
+            ProverResourceType::Gpu => {
+                #[cfg(feature = "cuda")]
+                {
+                    let $sdk = $zkvm.gpu_sdk()?;
+                    $f
+                }
+                #[cfg(not(feature = "cuda"))]
+                panic!("Feature `cuda` is disabled. Enable `cuda` to use GPU resource type");
+            }
+            ProverResourceType::Network(_) => {
+                panic!(
+                    "Network proving not yet implemented for OpenVM. Use CPU or GPU resource type."
+                );
+            }
+        }
+    };
 }
 
 impl zkVM for EreOpenVM {
     fn execute(
         &self,
         inputs: &Input,
     ) -> Result<(PublicValues, zkvm_interface::ProgramExecutionReport), zkVMError> {
-        let sdk = self
-            .sdk()
-            .map_err(|e| OpenVMError::from(ExecuteError::from(e)))?;
-
         let mut stdin = StdIn::default();
         serialize_inputs(&mut stdin, inputs);
 
         let start = Instant::now();
-        let public_values = sdk
-            .execute(self.app_exe.clone(), stdin)
+        let public_values = with_sdk!(self, |sdk| sdk.execute(self.app_exe.clone(), stdin))
```

Comment:
Do you know if there's any relevant time overhead in the SDK creation?
Mostly thinking if is it worth pulling the SDK creation to the ctor, and avoid that overhead (if any) in execute/prove/etc.

Also, not sure if the SDKs have a common trait -- if that's the case sounds nice since we could avoid this macro if we store the SDK as a field created in ctor.

One benefit is that the way things work now it will only panic if the CUDA feature wasn't provided on execute/prove/etc and not a soon as the ctor is called. But I don't think that's a big deal anyway since low chances it is a relevant detail.

(To be clear, not a strong opinion)